cdap-master/src/main/java/co/cask/cdap/data/tools/MetricsKafkaUpgrader.java

```diff
@@ -0,0 +1,155 @@
+import co.cask.cdap.data2.util.hbase.HBaseTableUtil;
+import co.cask.cdap.metrics.MetricsConstants;
+import co.cask.cdap.metrics.process.KafkaConsumerMetaTable;
+import co.cask.cdap.metrics.store.DefaultMetricDatasetFactory;
+import co.cask.cdap.proto.Id;
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableMap;
+import com.google.inject.Inject;
+import com.google.inject.name.Named;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.HConstants;
+import org.apache.hadoop.hbase.client.HTable;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.client.ResultScanner;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.twill.filesystem.LocationFactory;
+import org.apache.twill.kafka.client.TopicPartition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+/**
+ * Upgrade kafka metrics meta table data
+ */
+public class MetricsKafkaUpgrader extends AbstractUpgrader {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DatasetUpgrader.class);
+  private static final byte[] OFFSET_COLUMN = Bytes.toBytes("o");
+
+  private final Configuration hConf;
+  private final HBaseTableUtil hBaseTableUtil;
+  private final DatasetFramework dsFramework;
+  private final String kafkaTableName;
+  private final KafkaConsumerMetaTable kafkaMetaTableDestination;
+
+  @Inject
+  public MetricsKafkaUpgrader(CConfiguration cConf, Configuration hConf, LocationFactory locationFactory,
+                              HBaseTableUtil hBaseTableUtil, @Named("dsFramework") final DatasetFramework dsFramework) {
+    super(locationFactory);
+    this.hConf = hConf;
+    this.hBaseTableUtil = hBaseTableUtil;
+    this.dsFramework = dsFramework;
+    this.kafkaTableName =  cConf.get(MetricsConstants.ConfigKeys.KAFKA_META_TABLE,
+                                     MetricsConstants.DEFAULT_KAFKA_META_TABLE);
+    this.kafkaMetaTableDestination = new DefaultMetricDatasetFactory(cConf, dsFramework).createKafkaConsumerMeta();
+  }
+
+  private MetricsTable getOrCreateKafkaTable(String tableName, DatasetProperties props) {
+    MetricsTable table = null;
+    // old kafka table is in the default namespace
+    Id.DatasetInstance metricsDatasetInstanceId = Id.DatasetInstance.from(Constants.DEFAULT_NAMESPACE_ID, tableName);
+    try {
+      table = DatasetsUtil.getOrCreateDataset(dsFramework, metricsDatasetInstanceId,
+                                              MetricsTable.class.getName(), props, null, null);
+    } catch (Exception e) {
+      LOG.error("Exception while creating table {}.", tableName, e);
+    }
+    return table;
+  }
+
+  @Override
+  public void upgrade() throws Exception {
+    // copy kafka off set from old table to new kafka metrics table
+
+    String kafkaTableNameOld = Joiner.on(".").join(Constants.SYSTEM_NAMESPACE, kafkaTableName);
+    byte[] columnFamily = getColumnFamily(kafkaTableNameOld);
```

Comment:
do getHTable first - and go out if it is not there. Otherwise getColumnFamily will create a table if it is not there... We don't want that